shared-bindings/ota/__init__.c

```diff
@@ -0,0 +1,127 @@
+/*
+ * This file is part of the Micro Python project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2020 microDev
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "shared-bindings/ota/__init__.h"
+
+//| """OTA Module
+//|
+//| The `ota` module implements over-the-air update.
+//|
+//| There are two identical ota partitions ota_0/1, these
+//| contain different firmware versions.
+//| Having two partitions enables rollback functionality.
+//|
+//| The two partitions are defined as boot partition and
+//| next-update partition. Calling `ota.flash()` writes the
+//| next-update partition.
+//|
+//| After the next-update partition is written a validation
+//| check is performed and on a successful validation this
+//| partition is set as the boot partition. On next reset,
+//| firmware will be loaded from this partition.
+//|
+//| Here is the sequence of commands to follow:
+//|
+//| .. code-block:: python
+//|
+//|     import ota
+//|
+//|     ota.flash(buffer, offset)
+//|     ota.finish()
+//| """
+//| ...
+//|
+
+//| def switch() -> None:
+//|     """Switches the boot partition. On next reset,
+//|        firmware will be loaded from the partition
+//|        just switched over to."""
+//|     ...
+//|
+STATIC mp_obj_t ota_switch(void) {
+    common_hal_ota_switch();
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_0(ota_switch_obj, ota_switch);
+
+//| def finish() -> None:
+//|     """Validates flashed firmware, sets next boot partition.
+//|         **Must be called** after the firmware has been
+//|         completely written into the flash using `ota.flash()`.
+//|
+//|         This is to be called only once per update."""
```

Comment:
Why only once?

Reply:
Yes, I would make this idempotent.

Reply:
There is no harm in calling this multiple times after `ota` partition is completely written. 
I want users to not call this each time they flash a small chunk of the firmware.

The `finish()` function was not present in the initial api and was a part of `flash()`. 
Reason for factoring it out is because `flash()` is called multiple times when flashing small chunks and I don't have a way to figure out if the firmware is completely written.

Reply:
What harm is done by calling it after each chunk? I agree it might not be ideal but it should be ok.

Reply:
Calling `finish()` deinits the update and performs a validation check.

https://github.com/adafruit/circuitpython/blob/37ee5e683deb26ee691090d4f51b0ade4fa9b61c/ports/esp32s2/common-hal/ota/__init__.c#L123-L144

Reply:
What does `deinit` mean? Why can't we call it multiple times?

Reply:
I am finding the wording in docs a little vague. Following change should clear things.
```suggestion

```

Reply:
I did some moving around and found out a way to remove `finish()` altogether.